accepted/2021/experimental-package-shading/experimental-package-shading.md

```diff
@@ -0,0 +1,256 @@
+3. **Its runtime assets are renamed** such that they are specific to the referencing project. When the package created from that project is consumed as a package reference, the renamed runtime assets must not collide with copies of those assets from any other shaded or unshaded package reference to the original package.
+4. **References to its compile assets are not exposed** by the project that contains the shaded package reference. For example, public APIs in compile assets of that project must not use types from the shaded package's compile assets, as the shaded package's compile assets will not be available to projects that reference that project or the package created by packing it.
+
+We next consider how setting `Shade="true"` on a package reference will cause it to have these behaviors.
+
+### Asset renaming
+
+Asset renaming will be handled by restore-time shading. Setting `Shade="true"` on a package reference will cause restore-time shading to shade it. In the scope of this experiment, this is the only flag that will affect restore-time shading.
+
+### Non-transitivity
+
+NuGet's existing concept of [private assets](https://docs.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#controlling-dependency-assets) already provides a way to make a package reference non-transitive. If a package reference has all its assets set to be private assets, then the package reference will not be transitive.
+
+As producer-side shading requires the shaded package be private, setting `Shade="true"` on a package reference should implicitly set `PrivateAssets="all"` on that reference. Explicitly setting any other value should cause an error.
+
+### Asset repackaging
+
+> TODO: Figure out what to do here. Is there a better name for this?
+
+Copying runtime assets from a project's package reference into the package is currently nontrivial. It requires setting the [`CopyLocalLockFileAssemblies` property](https://docs.microsoft.com/en-us/dotnet/core/project-sdk/msbuild-props#copylocallockfileassemblies) to `True` so that private runtime assets are copied to the output directory, and then either manually adding [or writing an MSBuild target](https://github.com/mono/t4/blob/e2b0252ee958a4c1ec4394773f020b3e5e0b095f/Mono.TextTemplating.Build/Mono.TextTemplating.Build.csproj#L27) to do so.
+
+There is no existing metadata for a package reference that causes `Pack` to include its runtime assets in the package. We need a flag that does this, for example `BundleRuntimeAssets="true"` or `RepackAssets="runtime"`. We could expose this flag as an option for general purpose use, or keep it internal, but either way setting `Shade="true"` should enable it.
+
+### Private dependencies
+
+While a package _should_ not expose references to shaded compile assets, as consumers will be unable to resolve them, this does not have to be enforced by tooling. Within the scope of this experiment, we will leave it up to package authors.
```

Comment:
Is it possible for a package author to do this? Right now I think it would work like this:

1. The package doesn't expose the references
2. NuGet restores the primary assembly and the shaded assets into a folder
3. At build time, the SDK tasks read `project.assets.json` and nominates the primary assembly as a `@(Reference)`
4. RAR opens the primary assembly, sees that it references assemblies with shaded names, and tries to resolve them (looking next to the primary assembly)
5. If RAR finds them it exposes them to the compiler